File: tests/python/binary_search_trees/validate_three_nodes/test_validate_three_nodes.py
from problems.binary_search_trees.assets.BST import BST
from problems.binary_search_trees.validate_three_nodes.python.validate_three_nodes import \
    validateThreeNodes
from tests.python.binary_search_trees.utility import build_bst
from tests.python.utility import load_test_cases


def find_node(root: BST, targetValue: int) -> BST:
    if not root:
        return None
    if root.value == targetValue:
        return root
    nextNode = root.right if targetValue > root.value else root.left
    return find_node(nextNode, targetValue)


def test_validate_three_nodes() -> None:
    test_cases = load_test_cases("binary_search_trees/validate_three_nodes.json")

    for idx, test in enumerate(test_cases):
        tree = test["tree"]
        root = build_bst(tree["nodes"], tree["root"])

        nodeOne = find_node(root, int(test["nodeOne"]))
        nodeTwo = find_node(root, int(test["nodeTwo"]))
        nodeThree = find_node(root, int(test["nodeThree"]))

        expected = test["expected"]
        result = validateThreeNodes(nodeOne, nodeTwo, nodeThree)

        assert (
            result == expected
        ), f"Test case {idx} failed: expected {expected}, got {result}"

File: tests/cpp/binary_search_trees/test_validate_three_nodes.cpp
#include "../../../problems/binary_search_trees/validate_three_nodes/cpp/validate_three_nodes.h"
#include "../jsontestbase.h"
#include "./utility.h"

#include <sstream>

class ValidateThreeNodesTest : public JsonTestBase {
public:
  ValidateThreeNodesTest() {
    json_file_path = std::string(TEST_CASES_DIR) +
                     "/binary_search_trees/validate_three_nodes.json";
  }
};

BST *findNode(BST *root, int targetValue) {
  if (!root)
    return nullptr;

  if (root->value == targetValue)
    return root;

  BST *nextNode = (targetValue > root->value) ? root->right : root->left;
  return findNode(nextNode, targetValue);
}

int convertToInteger(const std::string &str) {
  std::stringstream ss(str);
  int value;
  ss >> value;
  return value;
}

TEST_F(ValidateThreeNodesTest, TestCases) {
  ASSERT_FALSE(test_cases.empty()) << "Test cases are empty or not loaded";

  for (const auto &test : test_cases) {
    auto treeJson = test["tree"];
    auto nodeOne = convertToInteger(test["nodeOne"]);
    auto nodeTwo = convertToInteger(test["nodeTwo"]);
    auto nodeThree = convertToInteger(test["nodeThree"]);
    auto expected = test["expected"];

    auto root = buildBST(treeJson);
    auto nodeOnePtr = findNode(root, nodeOne);
    auto nodeTwoPtr = findNode(root, nodeTwo);
    auto nodeThreePtr = findNode(root, nodeThree);

    auto result = validateThreeNodes(nodeOnePtr, nodeTwoPtr, nodeThreePtr);

    EXPECT_EQ(result, expected) << "Failed test case " << test.dump(2);

    deleteBST(root);
  }
}
